qiskit_dynamics/solver_classes.py

```diff
@@ -0,0 +1,335 @@
+        self.model.signals = new_signals
+
+    @property
+    def evaluation_mode(self) -> str:
+        """Evaluation mode for the model."""
+        return self.model.evaluation_mode
+
+    def set_evaluation_mode(self, new_mode: str):
+        """Set model evaluation mode. See documentation for
+        :class:`~qiskit_dynamics.models.HamiltonianModel`
+        or :class:`~qiskit_dynamics.models.LindbladModel`
+        (if dissipators present in model) for valid methods.
+        """
+        self.model.set_evaluation_mode(new_mode)
+
+    def copy(self) -> "Solver":
+        """Return a copy of self."""
+        return deepcopy(self)
+
+    def solve(
+        self, t_span: Array, y0: Union[Array, QuantumState, BaseOperator], **kwargs
+    ) -> OdeResult:
+        r"""Solve the dynamical problem using :func:`~qiskit_dynamics.solve_lmde`.
+
+        Special handling for various input types:
+            - If `y0` is an `Array`, it is passed directly to
+            :func:`~qiskit_dynamics.solve_lmde` as is. Acceptable array shapes are
+            determined by the model type and evaluation mode.
+            - If `y0` is a subclass of :class:`qiskit.quantum_info.QuantumState`:
+                - If `self.model` is a :class:`~qiskit_dynamics.models.LindbladModel`,
+                `y0` is converted to a :class:`DensityMatrix`. Further, if the model
+                evaluation mode is vectorized `y0` will be suitably reshaped for solving.
+                - If `self.model` is a :class:`~qiskit_dynamics.models.HamiltonianModel`,
+                and `y0` a :class:`DensityMatrix`, the full unitary will be simulated,
+                and the evolution of `y0` is attained via conjugation.
+            - If `y0` is a subclass of :class`qiskit.quantum_info.QuantumChannel`, the full
+            evolution map will be computed and composed with `y0`; either the unitary if
+            `self.model` is a :class:`~qiskit_dynamics.models.HamiltonianModel`, or the full
+            Lindbladian `SuperOp` if the model is a :class:`~qiskit_dynamics.models.LindbladModel`.
+
+        Returns an `OdeResult` object in the style of `scipy.integrate.solve_ivp`, with results
+        formatted to be the same types as the input.
+
+        Args:
+            t_span: Time interval to integrate over.
+            y0: Initial state.
+            kwargs: Keyword args passed to :meth:`~qiskit_dynamics.solve.solve_lmde`.
+
+        Returns:
+            OdeResult object with formatted output types.
+
+        Raises:
+            QiskitError: Initial state ``y0`` is of invalid shape.
+        """
+
+        # convert types
+        if isinstance(y0, QuantumState) and isinstance(self.model, LindbladModel):
+            y0 = DensityMatrix(y0)
+
+        y0, y0_cls = initial_state_converter(y0, return_class=True)
+
+        # validate types
+        if (
+            (y0_cls is SuperOp)
+            and isinstance(self.model, LindbladModel)
```

Comment:
This check and exception seems to get used a lot in various parts of the code so might be better to add a function like `check_vectorized_linbdlad_model(self.model)`
eg:
```python
def check_vectorized_lindblad_model(model):
    if isinstance(self.model, LindbladModel) and "vectorized" not in self.evaluation_mode:
        raise QiskitError(
                "Simulating SuperOp for a LinbladModel requires setting"
                " vectorized evaluation. Set evaluation_mode to a vectorized option.")
```

Reply:
Added two functions to `solvers/solver_utils.py` that check if an object is a LindbladModel and if it is vectorized v.s. not vectorized. Modified `solve.py` and `solver_classes.py` to use these functions.

Note: I wanted to put these in `type_utils.py` but it ended up causing a circular import, so decided to put them in `solvers/solver_utils.py`.